src/client_side_reply.cc

```diff
@@ -1838,8 +1836,7 @@ clientReplyContext::processReplyAccess ()
     /** Process http_reply_access lists */
     ACLFilledChecklist *replyChecklist =
         clientAclChecklistCreate(Config.accessList.reply, http);
-    replyChecklist->reply = reply;
-    HTTPMSGLOCK(replyChecklist->reply);
+    replyChecklist->updateReply(reply);
```

Comment:
Hmm, something seems wrong about that statement. These lines were correctly locking after setting the pointer.

Perhapse you meant that `clientAclChecklistCreate()` already set some value in reply that was not being correctly `HTTPMSGUNLOCK()`ed ?

Reply:
> These lines were correctly locking after setting the pointer.

No, they were not. As I tried to explain in the PR description, this code _looks_ OK, but it is buggy:

```C++
foo = bar;
HTTPMSGLOCK(foo); // BUG: Overlocking when foo was bar before the assignment
```

> Perhapse you meant that clientAclChecklistCreate() already set some value in reply that was not being correctly HTTPMSGUNLOCK()ed ?

I did not mean to say that (and did mean to say what I said). However, if we were stuck with manual locks/unlocks, then one could fix this bug by carefully unlocking replyChecklist->reply while resetting it to the same value and locking it again. The end result would be the same -- the final value of reference counter will decrease by one in the problematic case of self-assignment (it will stay the same/unchanged instead of being increased by one as in the current official code):

```C++
// one possible implementation that does not assume anything about prior values:
LOCK(bar); // in case foo is bar
UNLOCK(foo);
foo = bar;
```

Fortunately smart pointers do all this work for us.